sematic/ui/src/components/PipelineBar.tsx

```diff
@@ -1,69 +1,146 @@
 import RunStateChip from "./RunStateChip";
 import TimeAgo from "./TimeAgo";
+import { ActionMenu, ActionMenuItem } from "./ActionMenu";
+import { SnackBarContext } from "./SnackBarProvider";
+
+function PipelineActionMenu(props: {
+  rootRun: Run;
+  resolution: Resolution;
+  onCancel: () => void;
+}) {
+  const { rootRun, resolution, onCancel } = props;
+  const { user } = useContext(UserContext);
+  const theme = useTheme();
+  const { setSnackMessage } = useContext(SnackBarContext);
+
+  useEffect(() => {
+    pipelineSocket.removeAllListeners("cancel");
+    pipelineSocket.on("cancel", (args: { calculator_path: string }) => {
+      if (args.calculator_path === rootRun.calculator_path) {
+        setSnackMessage({ message: "Pipeline run was canceled." });
+        onCancel();
+      }
+    });
+  });
+
+  const onCancelClick = useCallback(() => {
+    if (!window.confirm("Are you sure you want to cancel this pipeline?"))
+      return;
+    fetchJSON({
+      url: "/api/v1/resolutions/" + rootRun.id + "/cancel",
+      method: "PUT",
+      apiKey: user?.api_key,
+      callback: (payload) => {},
+      setError: (error) => {
+        setSnackMessage({ message: "Failed to cancel pipeline run." });
+      },
+    });
+  }, [rootRun.id, setSnackMessage]);
+
+  const onRerunClick = useCallback(
+    (rerunFrom?: string) => {
+      fetchJSON({
+        url: "/api/v1/resolutions/" + rootRun.id + "/rerun",
+        method: "POST",
+        body: { rerun_from: rerunFrom },
+        apiKey: user?.api_key,
+        callback: (payload) => {},
+        setError: (error) => {
+          if (!error) return;
+          setSnackMessage({ message: "Failed to trigger rerun." });
+        },
+      });
+    },
+    [rootRun.id]
+  );
+
+  const onCopyShareClick = useCallback(() => {
+    navigator.clipboard.writeText(window.location.href);
+    setSnackMessage({ message: "Pipeline link copied." });
+  }, []);
+
+  const cancelEnabled = useMemo(
+    () =>
+      !["FAILED", "NESTED_FAILED", "RESOLVED", "CANCELED"].includes(
+        rootRun.future_state
+      ),
+    [rootRun.future_state]
+  );
+
+  const rerunEnable = useMemo(
+    () => !!resolution.container_image_uri,
+    [resolution.container_image_uri]
+  );
+
+  return (
+    <>
+      <ActionMenu title="Actions">
+        <ActionMenuItem
+          title="Rerun"
+          enabled={rerunEnable}
+          onClick={() => onRerunClick(rootRun.id)}
+        >
+          <Typography>Rerun pipeline from scratch.</Typography>
+          <Typography>Only available for remote resolutions.</Typography>
+        </ActionMenuItem>
+        <ActionMenuItem title="Retry from failure" enabled={false}>
+          <Typography>Rerun pipeline from where it failed.</Typography>
+          <Typography>Coming soon.</Typography>
+        </ActionMenuItem>
+        <ActionMenuItem title="Copy share link" onClick={onCopyShareClick}>
+          <Typography>Copy link to this exact pipeline execution.</Typography>
```

Comment:
I believe we should standardize the language and only use "resolution" (as above), and not use "pipeline execution".

Reply:
Done.